### ISSUE TITLE
Concept: Automatic export to GitHub Gists or other online document platform

### ISSUE THREAD

**Issue by ChrisAcrobat** [bitburner-official/bitburner-src#851]
**Idea**
1. In the option menu, add a login button to GitHub (or Google Docs, Dropbox, etc.). (EDIT: Steam?)
2. When logged in, game [creates a Gist](https://docs.github.com/en/rest/gists/gists?apiVersion=2022-11-28#create-a-gist) or [read](https://docs.github.com/en/free-pro-team@latest/rest/gists/gists?apiVersion=2022-11-28#get-a-gist) current save game and then [update it periodically](https://docs.github.com/en/rest/gists/gists?apiVersion=2022-11-28#update-a-gist).

Optionally add a profile or save game name, in case players would like to switch between saves.

**Comment by Hoekstraa** [comment]
What purpose would this be intended for? Backup? Distributed save file? It might be a smart idea to look at your root issue to discern the best solution.

**Comment by ChrisAcrobat** [comment]
Mostly distributed save file, but backup for hardware failure was also on my mind.

**Comment by Snarling** [comment]
Not planned at this time, if functionality like this is to be supported it would be via a generic save function and a load function.

The save function would provide a save string of the current save, which the player could then use to interact with a remote API if they wish. The load function would take in a save string. In either case, the interaction with some external service like github would be up to the player to set up within their own script.

But I'm not sure if save and load functions would really fit in to the API. They might be seen as making it too accessible for the player to modify the savegame. Maybe added as undocumented functions?

Made a poll on discord to get community feedback for this idea.

**Comment by catloversg** [comment]
We have not come to a conclusion in the discussion on Discord (https://discord.com/channels/415207508303544321/1207487024286801950/1207487117333106789). After rereading it, my suggestion is:
- Add a new documented API: `ns.getSaveData` or `ns.singularity.getSaveData`. It gives the player a Uint8Array. They can do whatever they want with it.
  - Should it be in the singularity namespace?
  - What does it cost? My suggestion: 0 GB if it's `ns.getSaveData`. 1GB if it's `ns.singularity.getSaveData`.
- Do not add any API to load the save data. It must be done manually via UI. If the player really wants to load the save data programmatically, they can use the webpack exploit to expose functions/objects in `SaveObject.ts` and use them.
- Keep `ns.singularity.exportGame` and `ns.singularity.exportGameBonus` as they are.

@Snarling @d0sboots I want to hear your opinions and the final decision.

**Comment by Snarling** [comment]
I think if we provide it, then it should probably be provided as a gzipped tar or whatever the save file is expected to be.

**Comment by catloversg** [comment]
That `Uint8Array` is the binary form of the save file (gz file). It's the same data as the thing downloaded from `ns.singularity.exportGame`. The difference is:
- New API: return the `Uint8Array` to the player. The player's script can use it.
- `ns.singularity.exportGame`: get the `Uint8Array` and trigger the download action.
  - On browser: download the save file (the content is that `Uint8Array`).
  - On Electron app: open the "Save as" popup to choose the location for downloading, then download the save file (the content is that `Uint8Array`).

By the way, where should we put that API? `ns` or `ns.singularity`?

**Comment by d0sboots** [comment]
I agree with those decisions. No opinion on `singularity` vs not; I can see arguments both ways.
